### ISSUE TITLE
False positive 'undefined-variable' with assignment expression in decorator

### ISSUE THREAD

**Issue by Jasha10** [PyCQA/pylint#5735]
### Bug description

When using an assignment expression in a decorator, a false positive 'undefined-variable' may be raised.

### Command used

```shell
# example.py
def preprocess(bar: str) -> str:
    return bar + "xyz"

def condition1(foo):
    return foo.startswith("abc")

def condition2(foo):
    return "bcdef" in foo

def postprocess(foo):
    return foo

data = ["abcdef", "qwerty"]

def decorator(arg):
    print(arg)
    return lambda x: x

@decorator(
    [postprocess(foo) for string in data if condition1(foo := preprocess(string)) and condition2(foo)],
)
def decorated() -> None:
    pass


`$ pylint example.py`
```

### Pylint output

```shell
************* Module example
tmp2.py:32:17: E0602: Undefined variable 'foo' (undefined-variable)
tmp2.py:32:97: E0602: Undefined variable 'foo' (undefined-variable)
tmp2.py:32:55: W0612: Unused variable 'foo' (unused-variable)

------------------------------------------------------------------
Your code has been rated at 3.12/10 (previous run: 3.12/10, +0.00)
```

### Expected behavior

No errors

### Pylint version

```shell
$ pylint --version
pylint 2.12.2
astroid 2.9.3
Python 3.9.9 | packaged by conda-forge | (main, Dec 20 2021, 02:40:17)
[GCC 9.4.0]
```

**Comment by DerBiasto** [comment]
I experienced similar behaviour in different contexts:

### Reproducible Example
```
print(x if (x := 1) else 0)
f = lambda: x if (x := 1) else 0


def g():
    return x if (x := 1) else 0
```

### Pylint output
```
pylint snippets/test_pylint.py --disable=C0114
************* Module test_pylint
snippets/test_pylint.py:1:6: E0601: Using variable 'x' before assignment (used-before-assignment)
snippets/test_pylint.py:2:12: E0602: Undefined variable 'x' (undefined-variable)
snippets/test_pylint.py:6:11: E0601: Using variable 'x' before assignment (used-before-assignment)

```
### Pylint version
```
pylint --version
pylint 2.7.2
astroid 2.5.1
Python 3.9.2 (default, Feb 28 2021, 17:03:44) 
[GCC 10.2.1 20210110]
```
